ui/summary.go

```diff
@@ -287,73 +361,78 @@ func SummarizeMetrics(w io.Writer, indent string, t time.Duration, timeUnit stri
-				markColor = FailColor
-			} else {
-				mark = SuccMark
-				markColor = SuccColor
-			}
-		}
+// Summarize summarizes a dataset in human readable format.
+func Summarize(w io.Writer, indent string, data SummaryData) {
+	if data.Root != nil {
+		summarizeGroup(w, indent+"    ", data.Root)
+	}
+	summarizeMetrics(w, indent+"  ", data.Time, data.Opts.SummaryTimeUnit.String, data.Metrics)
+}
 
-		fmtName := DisplayNameForMetric(m)
-		fmtIndent := IndentForMetric(m)
-		fmtName += GrayColor.Sprint(strings.Repeat(".", nameLenMax-StrWidth(fmtName)-StrWidth(fmtIndent)+3) + ":")
+func newJSONEncoder(w io.Writer) *json.Encoder {
+	encoder := json.NewEncoder(w)
+	encoder.SetIndent("", "    ")
+	return encoder
+}
+func summarizeGroupJSON(w io.Writer, group *lib.Group) {
+	encoder := newJSONEncoder(w)
+	if err := encoder.Encode(group); err != nil {
+		panic(err)
+	}
+}
 
-		var fmtData string
-		if cols := trendCols[name]; cols != nil {
-			for i, val := range cols {
-				tmpCols[i] = TrendColumns[i].Key + "=" + ValueColor.Sprint(val) + strings.Repeat(" ", trendColMaxLens[i]-StrWidth(val))
-			}
-			fmtData = strings.Join(tmpCols, " ")
-		} else {
-			value := values[name]
-			fmtData = ValueColor.Sprint(value) + strings.Repeat(" ", valueMaxLen-StrWidth(value))
-
-			extra := extras[name]
-			switch len(extra) {
-			case 0:
-			case 1:
-				fmtData = fmtData + " " + ExtraColor.Sprint(extra[0])
-			default:
-				parts := make([]string, len(extra))
-				for i, ex := range extra {
-					parts[i] = ExtraColor.Sprint(ex) + strings.Repeat(" ", extraMaxLens[i]-StrWidth(ex))
-				}
-				fmtData = fmtData + " " + ExtraColor.Sprint(strings.Join(parts, " "))
-			}
+func summarizeMetricsJSON(w io.Writer, t time.Duration, timeUnit string, metrics map[string]*stats.Metric) {
+	data := make(map[string]interface{})
+	for name, m := range metrics {
+		m.Sink.Calc()
+		if sink, ok := m.Sink.(*stats.TrendSink); ok {
+			data[name] = sink.Format(t)
+			continue
+		}
+		data[name] = m.Sink.Format(t)
+		_, extra := nonTrendMetricValueForSum(t, timeUnit, m)
+		if len(extra) > 1 {
+			extraData := make(map[string]interface{})
+			extraData["value"] = m.Sink.Format(t)["value"]
+			extraData["extra"] = extra
+			data[name] = extraData
 		}
-		_, _ = fmt.Fprint(w, indent+fmtIndent+markColor.Sprint(mark)+" "+fmtName+" "+fmtData+"\n")
+	}
+
+	encoder := newJSONEncoder(w)
+	if err := encoder.Encode(data); err != nil {
+		panic(err)
 	}
 }
 
-// Summarizes a dataset and returns whether the test run was considered a success.
-func Summarize(w io.Writer, indent string, data SummaryData) {
+// SummarizeJSON summarizes a dataset in JSON format.
+func SummarizeJSON(w io.Writer, data SummaryData) {
 	if data.Root != nil {
-		SummarizeGroup(w, indent+"    ", data.Root)
+		summarizeGroupJSON(w, data.Root)
 	}
-	SummarizeMetrics(w, indent+"  ", data.Time, data.Opts.SummaryTimeUnit.String, data.Metrics)
+	summarizeMetricsJSON(w, data.Time, data.Opts.SummaryTimeUnit.String, data.Metrics)
```

Comment:
Agree to return error instead of panic, fixed. But I think it's better to keep summary group and summary metrics separately. Just do one thing and do it well.